File: libraries/OclType.swift
import Foundation
import Darwin

class OclAttribute
{ private static var instance : OclAttribute? = nil

  init() { }

  init(copyFrom: OclAttribute) {
    self.name = copyFrom.name
    self.type = copyFrom.type
  }

  func copy() -> OclAttribute
  { let res : OclAttribute = OclAttribute(copyFrom: self)
    addOclAttribute(instance: res)
    return res
  }

  static func defaultInstance() -> OclAttribute
  { if (instance == nil)
    { instance = createOclAttribute() }
    return instance!
  }

  deinit
  { killOclAttribute(obj: self) }

  var name : String = ""
  var type : OclType = OclType.defaultInstance()

  func getName() -> String
  {
    return name
  }


  func getType() -> OclType
  {
    return type
  }

}


var OclAttribute_allInstances : [OclAttribute] = [OclAttribute]()

func createOclAttribute() -> OclAttribute
{ let result : OclAttribute = OclAttribute()
  OclAttribute_allInstances.append(result)
  return result
}

func addOclAttribute(instance : OclAttribute)
{ OclAttribute_allInstances.append(instance) }

func killOclAttribute(obj: OclAttribute)
{ OclAttribute_allInstances = OclAttribute_allInstances.filter{ $0 !== obj } }


class OclOperation
{ private static var instance : OclOperation? = nil

  init() { }

  init(copyFrom: OclOperation) {
    self.name = copyFrom.name
    self.type = copyFrom.type
    self.parameters = Ocl.copySequence(s: copyFrom.parameters)
  }

  func copy() -> OclOperation
  { let res : OclOperation = OclOperation(copyFrom: self)
    addOclOperation(instance: res)
    return res
  }

  static func defaultInstance() -> OclOperation
  { if (instance == nil)
    { instance = createOclOperation() }
    return instance!
  }

  deinit
  { killOclOperation(obj: self) }

  var name : String = ""
  var type : OclType = OclType.defaultInstance()
  var parameters : [OclAttribute] = [OclAttribute]()

  func getName() -> String
  {
    return name
  }


  func getType() -> OclType
  {
    return type
  }

  func getReturnType() -> OclType
  {
    return type
  }

  func addParameter(par: OclAttribute) 
  { parameters.append(par) } 

  func getParameters() -> [OclAttribute]
  { return parameters } 

}


var OclOperation_allInstances : [OclOperation] = [OclOperation]()

func createOclOperation() -> OclOperation
{ let result : OclOperation = OclOperation()
  OclOperation_allInstances.append(result)
  return result
}

func addOclOperation(instance : OclOperation)
{ OclOperation_allInstances.append(instance) }

func killOclOperation(obj: OclOperation)
{ OclOperation_allInstances = OclOperation_allInstances.filter{ $0 !== obj } }



class OclType
{ private static var instance : OclType? = nil

  init() { }

  init(copyFrom: OclType) {
    self.name = "copy_" + copyFrom.name
    self.attributes = Ocl.copySequence(s: copyFrom.attributes)
    self.operations = Ocl.copySequence(s: copyFrom.operations)
    self.constructors = Ocl.copySequence(s: copyFrom.constructors)
    self.innerClasses = Ocl.copySequence(s: copyFrom.innerClasses)
    self.componentType = Ocl.copySequence(s: copyFrom.componentType)
    self.superclasses = Ocl.copySequence(s: copyFrom.superclasses)
    self.subclasses = Ocl.copySequence(s: copyFrom.subclasses)
  }

  func copy() -> OclType
  { let res : OclType = OclType(copyFrom: self)
    addOclType(instance: res)
    return res
  }

  static func defaultInstance() -> OclType
  { if (instance == nil)
    { instance = createOclType() }
    return instance!
  }

  deinit
  { killOclType(obj: self) }

  var name : String = "" /* principal key */

  static var OclType_index : Dictionary<String,OclType> = [String:OclType]()

  static func getByPKOclType(index : String) -> OclType?
  { return OclType_index[index] }

  var attributes : [OclAttribute] = []
  var operations : [OclOperation] = []
  var constructors : [OclOperation] = []
  var innerClasses : [OclType] = []
  var componentType : [OclType] = []
  var superclasses : [OclType] = []
  var subclasses : [OclType] = []
  var actualMetatype : Any.Type? = nil

  func getName() -> String
  {
    return name
  }


  func getClasses() -> [OclType]
  { return innerClasses }


  func getDeclaredClasses() -> [OclType]
  {
    var result : [OclType] = []
    result = Ocl.sequenceSubtract(s1: innerClasses, s2: Ocl.unionAll(s: superclasses.map({sc in sc.getClasses()})))
    return result
  }


  func getComponentType() -> OclType
  {
    var result : OclType = OclType.defaultInstance()
    if componentType.count > 0
    {
      result = Ocl.any(s: componentType)!
    }
    else {
      if componentType.count == 0
      {
        result = createByPKOclType(key: "void")
      }
    }
    return result
  }


  func getFields() -> [OclAttribute]
  {
    return attributes
  }


  func getDeclaredField(s : String) -> OclAttribute
  {
    let result : OclAttribute = OclAttribute.defaultInstance()
    let x = Ocl.any(s: Ocl.select(s: attributes, f: { att in att.name == s }))
    if x != nil
    { return x! } 
    return result
  }


  func getField(s : String) -> OclAttribute
  {
    let result : OclAttribute = OclAttribute.defaultInstance()
    let x = Ocl.any(s: Ocl.select(s: attributes, f: { att in att.name == s }))
    if x != nil
    { return x! } 
    return result
  }


  func getDeclaredFields() -> [OclAttribute]
  {
    var result : [OclAttribute] = []
    result = Ocl.sequenceSubtract(s1: attributes, s2: Ocl.unionAll(s: superclasses.map({sc in sc.getFields()})))
    return result
  }


  func getMethods() -> [OclOperation]
  {
    return operations
  }


  func getDeclaredMethods() -> [OclOperation]
  {
    var result : [OclOperation] = []
    result = Ocl.sequenceSubtract(s1: operations, s2: Ocl.unionAll(s: superclasses.map({sc in sc.getMethods()})))
    return result
  }


  func getConstructors() -> [OclOperation]
  {
    return constructors
  }

  func addSuperclass(sup: OclType) 
  { superclasses.append(sup)
    sup.subclasses.append(self)
  } 

  func getSuperclass() -> OclType
  {
    var result : OclType? = nil
    if superclasses.count == 0
    { result = createByPKOclType(key: "OclAny")
      return result!
    } 
    result = Ocl.any(s: superclasses)
    return result!
  }

  static func hasAttribute(obj: Any, att: String) -> Bool
  { let m = Mirror(reflecting: obj)
    for e in m.children { 
      if e.label == att 
      { return true } 
    } 
    return false
  } 

  static func getAttributeValue(obj: Any, att: String) -> Any?
  { let m = Mirror(reflecting: obj)
    for e in m.children { 
      if e.label == att 
      { return e.value } 
    } 
    return nil
  } 

  func isArray() -> Bool
  { if name == "Sequence" 
    { return true } 
    return false
  } 

  func isPrimitive() -> Bool
  { if name == "int" || name == "long" || name == "double" || name == "boolean"  
    { return true } 
    return false
  } 

  func isAssignableFrom(c : OclType) -> Bool
  { if c.name == self.name 
    { return true } 
    for s in c.superclasses { 
      if isAssignableFrom(c: s)
      { return true } 
    } 
    return false
  } 

  func isInstance(obj : Any) -> Bool
  { if actualMetatype == type(of: obj)
    { return true } 
    for s in subclasses { 
      if s.isInstance(obj: obj)
      { return true } 
    } 
    return false
  } 

}


var OclType_allInstances : [OclType] = [OclType]()

func createOclType() -> OclType
{ let result : OclType = OclType()
  OclType_allInstances.append(result)
  return result
}

func addOclType(instance : OclType)
{ OclType_allInstances.append(instance) }

func killOclType(obj: OclType)
{ OclType_allInstances = OclType_allInstances.filter{ $0 !== obj } }

func createByPKOclType(key : String) -> OclType
{ var result : OclType? = OclType.getByPKOclType(index: key)
  if result != nil { return result! } 
  result = OclType()
  OclType_allInstances.append(result!)
  OclType.OclType_index[key] = result!
  result!.name = key
  return result! }

func killOclType(key : String)
{ OclType.OclType_index[key] = nil
  OclType_allInstances.removeAll(where: { $0.name == key })
}

File: libraries/OclType.java
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Constructor;
import java.util.Map; 
import java.util.HashMap;
import java.util.Collection;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;
import java.util.TreeSet;
import java.util.Collections;

class OclType { 
  static 
  { OclType.OclType_index = new HashMap<String,OclType>();
    OclType stringType = OclType.createByPKOclType("String"); 
    stringType.actualMetatype = String.class; 
    OclType intType = OclType.createByPKOclType("int"); 
    intType.actualMetatype = int.class; 
    OclType longType = OclType.createByPKOclType("long"); 
    longType.actualMetatype = long.class; 
    OclType doubleType = OclType.createByPKOclType("double"); 
    doubleType.actualMetatype = double.class; 
    OclType booleanType = OclType.createByPKOclType("boolean"); 
    booleanType.actualMetatype = boolean.class; 
    OclType voidType = OclType.createByPKOclType("void"); 
    voidType.actualMetatype = void.class; 
  } 

  public Class actualMetatype = null; 

  OclType() { }

  OclType(String nme)
  { name = nme; 
    OclType.createByPKOclType(nme); 
  } 

  OclType(Class c) 
  { this(c.getName());
    actualMetatype = c; 
  }


  static OclType createOclType() 
  { OclType result = new OclType();
    return result; 
  }

  String name = ""; /* primary */
  static Map<String,OclType> OclType_index;

  static OclType createByPKOclType(String namex)
  { OclType result = OclType.OclType_index.get(namex);
    if (result != null) 
    { return result; }
    result = new OclType();
    OclType.OclType_index.put(namex,result);
    result.name = namex;
    return result; }

  static void killOclType(String namex)
  { OclType rem = OclType_index.get(namex);
    if (rem == null) { return; }
    ArrayList<OclType> remd = new ArrayList<OclType>();
    remd.add(rem);
    OclType_index.remove(namex);
  }

  ArrayList<OclAttribute> attributes = new ArrayList<OclAttribute>();
  ArrayList<OclOperation> operations = new ArrayList<OclOperation>();
  ArrayList<OclOperation> constructors = new ArrayList<OclOperation>();
  ArrayList<OclType> innerClasses = new ArrayList<OclType>();
  ArrayList<OclType> componentType = new ArrayList<OclType>();
  ArrayList<OclType> superclasses = new ArrayList<OclType>();

  public void setMetatype(Class cls)
  { actualMetatype = cls; } 

  public String getName()
  { return name; }


  public ArrayList<OclType> getClasses()
  { return innerClasses; }


  public ArrayList<OclType> getDeclaredClasses()
  {
    ArrayList<OclType> result = new ArrayList<OclType>();
    result = Ocl.subtract(innerClasses,Ocl.unionAll(Ocl.collectSequence(superclasses,(sc)->{return sc.getClasses();})));
    return result;
  }


  public OclType getComponentType()
  {
    OclType result = null;
    if (componentType.size() > 0)
    {
      result = Ocl.any(componentType);
    }
    return result;
  }


  public ArrayList<OclAttribute> getFields()
  { if (actualMetatype != null) 
    { attributes.clear(); 
      Field[] flds = actualMetatype.getFields(); 
      for (int i = 0; i < flds.length; i++) 
      { OclAttribute att = new OclAttribute(flds[i].getName()); 
        att.setType(new OclType(flds[i].getType())); 
        attributes.add(att); 
      } 
    } 
    return attributes;
  }

  public static ArrayList<OclAttribute> allAttributes(Object obj)
  { ArrayList<OclAttribute> res = new ArrayList<OclAttribute>(); 
    if (obj == null) 
    { return res; } 

    if (obj instanceof Map)
    { Map mp = (Map) obj; 

      java.util.Set keys = mp.keySet(); 
      for (Object k : keys)
      { String key = k + ""; 
        OclAttribute katt = 
           new OclAttribute(key);
        Object val = mp.get(k);
        if (val != null) 
        { Class vtype = val.getClass(); 
          OclType otype = new OclType(vtype); 
          katt.setType(otype); 
        } 
        res.add(katt); 
      } 
      return res; 
    } 
      
    Class metatype = obj.getClass(); 
    Field[] flds = metatype.getFields(); 
    for (int i = 0; i < flds.length; i++) 
    { OclAttribute att = new OclAttribute(flds[i].getName()); 
      att.setType(new OclType(flds[i].getType())); 
      res.add(att); 
    }
    return res;  
  } 


  public OclAttribute getDeclaredField(String s)
  { attributes = getFields(); 
    OclAttribute result = null;
    result = Ocl.any(Ocl.selectSequence(attributes,(att)->{return att.name.equals(s);}));
    return result;
  }


  public OclAttribute getField(String s)
  { attributes = getFields(); 
    OclAttribute result = null;
    result = Ocl.any(Ocl.selectSequence(attributes,(att)->{return att.name.equals(s);}));
    return result;
  }


  public ArrayList<OclAttribute> getDeclaredFields()
  {
    ArrayList<OclAttribute> result = new ArrayList<OclAttribute>();
    if (actualMetatype != null) 
    { Field[] flds = actualMetatype.getDeclaredFields(); 
      for (int i = 0; i < flds.length; i++) 
      { OclAttribute att = new OclAttribute(flds[i].getName()); 
        att.setType(new OclType(flds[i].getType())); 
        result.add(att); 
      } 
    } 
    return result;
  }


  public ArrayList<OclOperation> getMethods()
  { if (actualMetatype != null) 
    { operations.clear(); 
      Method[] mets = actualMetatype.getMethods(); 
      for (int i = 0; i < mets.length; i++) 
      { OclOperation op = new OclOperation(mets[i].getName()); 
        if (mets[i].getReturnType() != null) 
        { op.setType(new OclType(mets[i].getReturnType())); }  
        operations.add(op); 
      } 
    } 
    return operations; 
  }


  public ArrayList<OclOperation> getDeclaredMethods()
  {
    ArrayList<OclOperation> result = new ArrayList<OclOperation>();
    if (actualMetatype != null) 
    { Method[] mets = actualMetatype.getDeclaredMethods(); 
      for (int i = 0; i < mets.length; i++) 
      { OclOperation op = new OclOperation(mets[i].getName()); 
        if (mets[i].getReturnType() != null) 
        { op.setType(new OclType(mets[i].getReturnType())); }  
        result.add(op); 
      } 
    } 
    return result;
  }


  public ArrayList<OclOperation> getConstructors()
  {
    ArrayList<OclOperation> result = new ArrayList<OclOperation>();
    if (actualMetatype != null) 
    { Constructor[] mets = actualMetatype.getDeclaredConstructors(); 
      for (int i = 0; i < mets.length; i++) 
      { OclOperation op = new OclOperation(mets[i].getName()); 
        result.add(op); 
      } 
    } 
    return result;
  }


  public OclType getSuperclass()
  { if (actualMetatype != null) 
    { Class superclass = actualMetatype.getSuperclass();
      if (superclass != null)
      { superclasses.clear(); 
        superclasses.add(new OclType(superclass)); 
      } 
    } 
    OclType result = null;
    if (superclasses.size() > 0)
    { result = Ocl.any(superclasses); } 
    return result;
  }

  public static boolean hasAttribute(Object obj, String att)
  { if (obj == null) 
    { return false; } 

    if (obj instanceof Map)
    { Map mp = (Map) obj; 
      java.util.Set keys = mp.keySet(); 
      for (Object k : keys)
      { if (att.equals(k + ""))
        { return true; } 
      } 
      return false; 
    } 

    Class metatype = obj.getClass(); 
    try {
      Field fld = metatype.getField(att); 
      return fld != null;
    } catch (Exception e) 
      { return false; }  
  } 

  public static Object getAttributeValue(Object obj, String att)
  { if (obj == null) 
    { return null; } 

    if (obj instanceof Map)
    { Map mp = (Map) obj; 
      java.util.Set keys = mp.keySet(); 
      for (Object k : keys)
      { if (att.equals(k + ""))
        { return mp.get(k); } 
      } 
      return null; 
    } 

    Class metatype = obj.getClass(); 
    try {
      Field fld = metatype.getField(att); 
      if (fld != null)
      { return fld.get(obj); } 
    } catch (Exception e) 
      { }  
    return null; 
  } 

  public static void setAttributeValue(Object obj, String att, Object val)
  { if (obj == null) 
    { return; } 

    if (obj instanceof Map)
    { Map mp = (Map) obj; 
      mp.put(att,val); 
      return; 
    } 

    Class metatype = obj.getClass(); 
    try {
      Field fld = metatype.getField(att); 
      if (fld != null)
      { fld.set(obj,val); } 
    } catch (Exception e) 
      { }  
  } 

  public static void removeAttribute(Object obj, String att)
  { if (obj instanceof Map)
    { Map mp = (Map) obj; 
      mp.remove(att); 
      return; 
    } 

    System.err.println(
       "removeAttribute is invalid in Java");
  }  
  
  public Object newInstance()
  { if (actualMetatype != null) 
    { try { return actualMetatype.newInstance(); }
	  catch (InstantiationException _e) { return null; }
	  catch (IllegalAccessException _e) { return null; }
    } 
	return null; 
  } 

  public boolean isArray()
  { if ("Sequence".equals(name)) 
    { return true; } 

    if (actualMetatype != null)
    { return actualMetatype.isArray(); } 

    return false; 
  }  

  public boolean isPrimitive()
  { if ("int".equals(name) || 
        "long".equals(name) ||
        "double".equals(name) ||
        "boolean".equals(name)) 
    { return true; } 

    if (actualMetatype != null)
    { return actualMetatype.isPrimitive(); } 

    return false; 
  }  

  public boolean isInterface()
  { if (actualMetatype != null) 
    { return actualMetatype.isInterface(); } 
    return false; 
  } 

  public boolean isAssignableFrom(OclType c)
  { if (c.actualMetatype != null && actualMetatype != null) 
    { return actualMetatype.isAssignableFrom(
                                 c.actualMetatype); 
    } 
    return false; 
  } 

  public boolean isInstance(Object obj)
  { if (actualMetatype != null) 
    { return actualMetatype.isInstance(obj); } 
    return false; 
  } 
  
  public static OclType loadExecutableObject(String s)
  { try { 
      Class c = Class.forName(s); 
      if (c != null) 
	  { return new OclType(c); }
    } 
    catch (Exception _ex) { return null; }
    return null; 
  }

  public static void unloadExecutableObject(String s)
  { /* Not supported in Java */  
  }
}

